lib/liquid/standardfilters.rb

```diff
@@ -213,17 +213,23 @@ def where(input, property, target_value = nil)
 
       if ary.empty?
         []
-      elsif ary.first.respond_to?(:[]) && target_value.nil?
-        begin
-          ary.select { |item| item[property] }
+      elsif target_value.nil?
+        ary.select do |item|
+          item[property]
         rescue TypeError
           raise_property_error(property)
+        rescue NoMethodError
+          return nil unless item.respond_to?(:[])
+          raise
```

Comment:
First of all, I think you meant `if` instead of `unless`.

Second of all, it would behave differently.  For instance, the `assert_nil(@filters.where([nil], "ok"))` assertion in `test_where_array_of_only_unindexable_values` would fail with `Expected [] to be nil.` since we were returning `nil` for the whole method not just the block when an item doesn't respond to `[]`.

Lastly, doing that check there would add overhead for each item.  By handling `NoMethodError` we avoid that extra overhead on the fast path, by leveraging ruby's method lookup to do that check for us.

Reply:
> doing that check there would add overhead for each item

Ah you're totally right!